---
title: "Minda Barcoding"
subtitle: '**---Analysis Documentation---**'
author: |
  | Kevin L. Labrador
  |
  | CRREST Laboratory
  | Texas A&M University - Corpus Christi
date: "`r Sys.Date()`"
output: 
  bookdown::html_document2:
    toc: true
    number_sections: true
    df_print: tibble
    fig_caption: true
csl: https://gist.githubusercontent.com/dangxdong/1861117/raw/9f12c6ae6b1ffdc94d3132934ed6edc6893f5a98/molecular-ecology.csl
link-citations: yes
linkcolor: blue
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE
                      , results = 'hide'
                      , warning = FALSE
                      , message = FALSE
                      , fig.height = 5
                      , fig.width = 5
)
```

# Introduction
This documents the analyses done on the Minda barcoding paper (title pending).

## Housekeeping

Housekeeping section cleans up the global environment, prepares the working directories, and call the package, **pacman**.

```{r housekeeping}
# Clear the global environment.
rm(list=ls())

# Set-up the working directory in the source file location: 
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

wd <- getwd() # Do this after the working directory was set to source file location

# Assign directories to objects in the global environment
indir <- paste0("../INDIR/")
outdir <- paste0("../OUTDIR/")
scripts <- paste0("../scripts/")

# Check content of indir using list.files()
list.files(indir)
list.files(scripts)
```
## Loading libraries
Load the libraries needed for this pipeline.

```{r}
pacman::p_load (ape, tidyverse, janitor)
```

## Load data files

Identify all the infiles for this session.
```{r}
# Fasta file of the alignment
## Make sure that the alignment is done outside R.
infile_fasta <- paste0(indir, 
                       "BOLD-output/20230902_minda-barcoding alignment.fasta")

infile_csv <- list.files(paste0(indir, 
                                "BOLD-output"), 
                         pattern = "\\.csv$",
                         full.names = T)

infile_csv_filenames <- list.files(paste0(indir, 
                                "BOLD-output"), 
                         pattern = "\\.csv$")
```

Load the files
```{r}
aln.fas <- read.FASTA(infile_fasta,
                      type = "DNA")

files.csv <- lapply (infile_csv, 
                     read.csv, 
                     na.strings = "N/A") %>% 
  map(., ~clean_names(.x))

names(files.csv) <- infile_csv_filenames

```


Extract metadata from the alignment file
```{r}
aln.metadata <- data.frame (id = names(aln.fas)) %>% 
  separate_wider_delim(id,
                       delim = "|",
                       names = c("process_id", "taxa_id", "sample_id", "bin", NA),
                       cols_remove = T
  ) %>% 
  separate_wider_delim(taxa_id,
                       delim = "_",
                       names = c("genus", "species"),
                       too_few = "align_start",
                       cols_remove = F
  )



```

Distance Summaries

```{r}
species.dist <- files.csv$`divergence_within-species.csv` %>% 
  mutate (level = "species")

genus.dist <- files.csv$`divergence_wthin-genus.csv` %>% 
  mutate (level = "genus")

family.dist <- files.csv$`divergence_within-family.csv` %>% 
  mutate (level = "family")

dist.df <- rbind (species.dist,
                  genus.dist,
                  family.dist) %>% 
  mutate (level = fct_relevel(level, "species", "genus", "family"))

```

```{r}
plot_distance_histogram <- 
    ggplot(dist.df, aes(x = distance, fill = level)) + 
    geom_histogram(col="black", 
                   position = "dodge") +
    coord_cartesian(expand = F) +
    scale_x_continuous(breaks = seq(0, 30, by=2)) +
    labs(x = "K2P Genetic Distance", 
         y = "Frequency",
         fill = "Taxonomic Level"
    ) + 
    theme_classic() +
  scale_fill_brewer(palette = "Accent")
plot_distance_histogram  

ggsave(plot_distance_histogram,
       file = paste0(outdir,
                     "K2P-distances_histogram.png"),
       height = 3.5,
       width = 6,
       units = "in",
       dpi = 330
)
```

```{r}
plot_distance_boxplot <- 
  ggplot(dist.df, aes(x = level, y = distance, fill = level)) + 
  geom_boxplot(col="black", 
               position = "dodge") +
  labs(x = "Taxonomic Level", 
       y = "K2P Genetic Distance") + 
  guides (fill = "none") +
  scale_fill_brewer(palette = "Accent") +
  theme_classic()
plot_distance_boxplot

ggsave(plot_distance_boxplot,
       file = paste0(outdir,
                     "K2P-distances_boxplot.png"),
       height = 3.5,
       width = 3.5,
       units = "in",
       dpi = 330
)
```



