---
title: "Minda Barcoding"
subtitle: '**---Analysis Documentation---**'
author: |
  | Kevin L. Labrador
  |
  | CRREST Laboratory
  | Texas A&M University - Corpus Christi
date: "`r Sys.Date()`"
output: 
  bookdown::html_document2:
    toc: true
    number_sections: true
    df_print: tibble
    fig_caption: true
csl: https://gist.githubusercontent.com/dangxdong/1861117/raw/9f12c6ae6b1ffdc94d3132934ed6edc6893f5a98/molecular-ecology.csl
link-citations: yes
linkcolor: blue
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE
                      , results = 'hide'
                      , warning = FALSE
                      , message = FALSE
                      , fig.height = 5
                      , fig.width = 5
)
```

# Introduction
This 


Phylogenetic analysis based on R. Follows after the analysis pipeline on [Ph-IRES](https://github.com/Ph-IRES/workshop_data-analysis/tree/main/tutorial_phylogeny_analysis).

## Before you start

This pipeline assumes that the sequence has already been aligned. As much as possible, gaps/indels are removed from the alignment so as to avoid errors downstream. Furthermore, ambiguous sequences should have been resolved prior to uploading the data in R.

There are freely-available software that can guide in the upstream processes (e.g., sequence truncation and clean-up, alignment). @Hall2018 is a good reference for sequence alignment, as well as other topics involving phylogenetics.

## Housekeeping

Housekeeping section cleans up the global environment, prepares the working directories, and call the package, **pacman**.

```{r housekeeping}
# Clear the global environment.
rm(list=ls())

# Set-up the working directory in the source file location: 
#setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

wd <- getwd() # Do this after the working directory was set to source file location

# Create (1) INPUT and (2) OUTPUT directories. 
# All input to and output from R should be located in these directories.
dir.create ("INDIR")
dir.create ("OUTDIR")

# Assign directories to objects in the global environment
indir <- paste0(wd, "/INDIR/")
outdir <- paste0(wd, "/OUTDIR/")

# Check content of indir using list.files()
list.files(indir)
```

### Loading libraries

Use package manager (`pacman`) in loading packages. `pacman` will load packages; in case a package is not available, it will automatically download the package from CRAN.

```{r library}
library(pacman)

# Load packages
pacman::p_load(
  
  # Wrangling, Miscellaneous
  tidyverse,
  janitor,

  # Plotting
  ggplot2, 
  ggpubr, 
  ggtree, 
  paletteer, 
  patchwork, 
  scales, 
  
  # Phylogenetics and Population Genetics Packages
  ade4, 
  adegenet, 
  ape, 
  bioseq,
  haplotypes, 
  pegas, 
  phangorn, 
  msa
)

```

# Initialization

## Load alignment file

Load the alignment file in fasta format.

```{r input-fasta}
infile <- list.files (indir,
                      pattern = "fasta",
                      full.names = T)

fasta_file <- read.FASTA(infile,
                        type = "DNA")
```

## Load metadata file

Load the metadata file in csv format. This should be modified based on how you would want to design trees later.

```{r input-metadata}
infile <- list.files (indir,
                      pattern = "csv",
                      full.names = T)

metadata <- read.csv (infile) %>% 
  clean_names() %>% 
  mutate (organism = str_replace_all(string = organism,
                                     pattern = "\\s+", 
                                     replacement = "_"),
          
          tree_label = paste(organism,
                             name, 
                             sep=":") 
  ) %>% 
  dplyr::rename (label = name)
```

Create a data back-up containing the alignment (fasta) and metadata (csv) information.

```{r create-bkp}
data_bkp <- as_tibble.DNAbin(fasta_file) %>% 
  full_join(metadata)

```

## Modify labels

Change the labels of the fasta file based on the information from the metadata file.

```{r change_aln-label}
# Select only the columns needed
phylo_aln <- data_bkp %>% 
  select (tree_label, sequence) %>% 
  as_DNAbin (labels = tree_label,
             sequences = sequence)

```

# Collapse Haplotypes

```{r collapse-haplotypes}
haplo_list <- phylo_aln %>% 
  as.dna %>% 
  haplotypes::haplotype(., indels="sic")

# Haplotype report
sink (paste(outdir, "hapCollapse_report.txt"))
print (haplo_list)
sink()
  
# Set filters
filter <- map(haplo_list@haplist,
    ~.x[1]) %>% 
  plyr::ldply() %>% 
  pull (V1)

# Collapse haplotypes based on filter
phylo_haplo_aln <- phylo_aln[filter]

```

# Model Test

Determine the best evolutionary model for the given dataset using Maximum Likelihood.

```{r ML Model Selection}
# Conver collapsed alignment into a PhyDat object
pd<-as.phyDat(phylo_haplo_aln)

##test models, select lowest AIC
mt <- modelTest(pd, multicore = TRUE, mc.cores= 6)
View(mt)
```

# Maximum Likelihood

We now infer the phylogeny using maximum likelihood, and then visualize the resulting phylogenetic tree.

## Model Initialization

Initialize the model fit for the dataset.

```{r Initial ModelFit}
##initial fit, use best fit model from prev chunk
fit <- as.pml(mt)
fit
```

## Bootstrapping

Optimize the fit of the model and perform bootstrapping.

```{r Optimize Fit and Bootstrap}
##MAKE SURE TO CHANGE OPTIONS DEPENDING ON YOUR SELECTED MODEL.  
#?optim.pml gives guidance on how to set optBf and optQ
#optInv and optGamma should be set depending on whether your model includes +I and/or +G parameters

fit.opt <- optim.pml(fit, 
                     optBf=TRUE, 
                     optQ=TRUE, 
                     optInv=TRUE, 
                     optGamma=TRUE,
                     rearrangement = "NNI", 
                     control = pml.control(trace = 1)
                     )
fit.opt

#main= sets the plot title
plot(fit.opt, 
     main="GTR+G(4)+I")

#bootstrap model
bs <- bootstrap.pml(fit.opt, 
                    bs=1000, 
                    optNni=TRUE, 
                    multicore=TRUE, 
                    mc.cores = 6)

#Save bs in case as an R data file
# write_rds(bs,
#           file = paste0(outdir, "bootstrap.tre")
#           )

# In case you want to load the bs file:
# bs <- readRDS(paste0(outdir, "bootstrap.tre"))
          
```

## Initial Visualization

Look at the initial plot of the tree.

```{r}
tree<-plotBS(fit.opt$tree, bs, p = 50, digits = 0, type="p", method="FBP")

#digits doesn't seem to work in plotBS, so this is a workaround
tree$node.label<-round(tree$node.label, digits=0)

# plotBS also saves the tree and the bootstrap values in a phylo object, which we will use for visualization using ggtree.
```

```{r}
# Get a basic view of the tree
ggtree (tree,
        ladderize = T) +
  xlim (NA, 1) +
  geom_tiplab(geom="text", fontface=3, size=2.5, hjust = -0.025) +
  geom_text(aes(label=node), hjust=0.5, size = 2)


```

```{r}
# Set outgroup node
out.node <- 45
tree.root <- root(tree, node=out.node)

# Review tree after rooting
(treeplot <- 
ggtree (tree.root,
        ladderize = T) +
  xlim (NA, 1) +
  geom_tiplab(geom="text", fontface=3, size=3.5, hjust = -0.025) +
  geom_nodelab(size=3, hjust = 1.25, vjust=-0.2))

```

```{r modify and manipulate trees}

# Check node numbers
treeplot + geom_text(aes(label=node), hjust=0.5, size = 2)


#hilight nodes
treeplot_2 <- 
treeplot + 
  geom_hilight(node = 1, fill="darkred", extend = 0.55, alpha=0.25) +
  geom_treescale(width = 0.1, fontsize = 3, offset=-1)


```

## Exporting phylogeny

```{r tree tools}
##export tree in Newick format
write.tree(tree, file=paste0(outdir, "newick.tree"))

##export tree if png format
ggsave (treeplot_2,
        file = paste0(outdir, "phylo_tree.png"),
        width = 5,
        height = 7,
        units = "in")



```
