---
title: "Minda Barcoding"
subtitle: '**---Analysis Documentation---**'
author: |
  | Kevin L. Labrador
  |
  | CRREST Laboratory
  | Texas A&M University - Corpus Christi
date: "`r Sys.Date()`"
output: 
  bookdown::html_document2:
    toc: true
    number_sections: true
    df_print: tibble
    fig_caption: true
csl: https://gist.githubusercontent.com/dangxdong/1861117/raw/9f12c6ae6b1ffdc94d3132934ed6edc6893f5a98/molecular-ecology.csl
link-citations: yes
linkcolor: blue
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE
                      , results = 'hide'
                      , warning = FALSE
                      , message = FALSE
                      , fig.height = 5
                      , fig.width = 5
)
```

# Introduction
This documents the analyses done on the Minda barcoding paper (title pending).

## Housekeeping

Housekeeping section cleans up the global environment, prepares the working directories, and call the package, **pacman**.

```{r housekeeping}
# Clear the global environment.
rm(list=ls())

# Set-up the working directory in the source file location: 
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

wd <- getwd() # Do this after the working directory was set to source file location

# Assign directories to objects in the global environment
indir <- paste0("../INDIR/")
outdir <- paste0("../OUTDIR/")
scripts <- paste0("../scripts/")

# Check content of indir using list.files()
list.files(indir)
list.files(scripts)
```
## Loading libraries
Load the libraries needed for this pipeline.

```{r}
pacman::p_load (ape, cowplot, tidyverse, janitor)
```

## Load data files

Identify all the infiles for this session.
```{r}
# Fasta file of the alignment
## Make sure that the alignment is done outside R.
infile_fasta <- paste0(indir, 
                       "BOLD-output/20230902_minda-barcoding alignment.fasta")

infile_csv <- list.files(paste0(indir, 
                                "BOLD-output"), 
                         pattern = "\\.csv$",
                         full.names = T)

infile_csv_filenames <- list.files(paste0(indir, 
                                "BOLD-output"), 
                         pattern = "\\.csv$")
```

Load the files
```{r}
aln.fas <- read.FASTA(infile_fasta,
                      type = "DNA")

files.csv <- lapply (infile_csv, 
                     read.csv, 
                     na.strings = "N/A") %>% 
  map(., ~clean_names(.x))

names(files.csv) <- infile_csv_filenames

```


Extract metadata from the alignment file
```{r}
aln.metadata <- data.frame (label = names(aln.fas)) %>% 
  separate_wider_delim(label,
                       delim = "|",
                       names = c("process_id", "taxa_id", "sample_id", "bin", NA),
                       cols_remove = F
  ) %>% 
  separate_wider_delim(taxa_id,
                       delim = "_",
                       names = c("genus", "species"),
                       too_few = "align_start",
                       cols_remove = F
  )



```

Create a data back-up containing the alignment (fasta) and metadata (csv) information.

```{r create-bkp}

data.bkp <- as_tibble.DNAbin(aln.fas) %>% 
  full_join(aln.metadata)

```

# Qualitative Data Description

## Diversity Profile

```{r}
taxo_info <- files.csv$`project-metadata_taxonomy.csv` 
```

```{r}
family_count <- taxo_info %>% 
  distinct (family, species) %>% 
  group_by (family) %>% 
  summarize (count = n() ) %>% 
  ungroup()
```

```{r}
plot_family_count <- 
ggplot (family_count,
        aes (x = reorder(family, count, decreasing = T),
             y = count)
        ) +
  geom_col(fill = "gray80",
           col = "black",
           position = "dodge") + 
  theme_classic()  +
  labs (x = "Family",
        y = "Number of species") +
  scale_y_continuous(expand = c(0,0)) +
  theme(axis.text.x = element_text(angle=90, 
                                   vjust = 0.25, 
                                   hjust = 1,
                                   size = 7)) 
plot_family_count

ggsave(plot_family_count,
       file = paste0(outdir,
                     "family-count_barplot.png"),
       height = 5,
       width = 7,
       units = "in",
       dpi = 330
)
```


## Distance Summaries

### Data Preparation

```{r}
species.dist <- files.csv$`divergence_within-species.csv` %>% 
  mutate (level = "species")

genus.dist <- files.csv$`divergence_wthin-genus.csv` %>% 
  mutate (level = "genus")

family.dist <- files.csv$`divergence_within-family.csv` %>% 
  mutate (level = "family")

dist.df <- rbind (species.dist,
                  genus.dist,
                  family.dist) %>% 
  mutate (level = fct_relevel(level, "species", "genus", "family"))

```

### Histogram
```{r}
plot_distance_histogram <- 
    ggplot(dist.df, aes(x = distance, fill = level)) + 
    geom_histogram(position = "dodge",
                   col = "black",
                   size = 0.25
                   ) +
    coord_cartesian(expand = F) +
    scale_x_continuous(breaks = seq(0, 30, by=2)) +
    labs(x = "K2P Genetic Distance (%)", 
         y = "Frequency",
         fill = "Taxonomic\nLevel"
    ) + 
    theme_classic() +
  scale_fill_brewer(palette = "Set1") 
plot_distance_histogram  

ggsave(plot_distance_histogram,
       file = paste0(outdir,
                     "K2P-distances_histogram.png"),
       height = 3.5,
       width = 6,
       units = "in",
       dpi = 330
)
```
### Boxplot
```{r}
plot_distance_boxplot <- 
  ggplot(dist.df, aes(x = level, y = distance, fill = level)) + 
  geom_boxplot(#col="black", 
               position = "dodge") +
  labs(x = "Taxonomic Level", 
       y = "K2P Genetic Distance (%)") + 
  guides (fill = "none") +
  scale_fill_brewer(palette = "Set1") +
  theme_classic()
plot_distance_boxplot

ggsave(plot_distance_boxplot,
       file = paste0(outdir,
                     "K2P-distances_boxplot.png"),
       height = 3.5,
       width = 3.5,
       units = "in",
       dpi = 330
)
```


## Barcode Gap

### Histogram

```{r}
barcode_gap <- files.csv$barcode_gap_analysis.csv %>% 
  pivot_longer(cols = c(mean_intra_sp,max_intra_sp, distance_to_nn),
               names_to = "comp",
               values_to = "dist"
  )

plot_barcode_gap_histogram <- 
ggplot (data = barcode_gap %>% filter (comp != "mean_intra_sp"), 
        aes (x = dist, 
             fill = comp)
) +
  geom_histogram(position="dodge",
                 col = "black",
                 size = 0.25
                 ) +
  coord_cartesian(expand = F) +
  theme_classic() +
  labs (x = "K2P Genetic Distance (%)", 
        y = "Frequency",
        fill = NULL) +
  scale_fill_brewer(palette = "Set1",
                    labels = c("interspecific\n(NN)", "intraspecific\n(max)")
  ) +
  theme(legend.text = element_text(size = 7)) 
plot_barcode_gap_histogram

```


### Scatterplot
```{r}
barcode_gap <- files.csv$barcode_gap_analysis.csv

plot_barcode_gap_scatterplot <- 
ggplot (data = barcode_gap,
        aes (x = max_intra_sp,
             y = distance_to_nn)
) +
  geom_jitter(pch=21,
             fill = "black",
             alpha = 0.50,
             col = "black",
             size = 1.25) +
  theme_classic() +
  coord_cartesian(expand = F) +
  lims (x = c(-0.5,30), 
        y = c(-0.5,30)
        ) +
  labs (x = "Maximum Intra-specific Distance (%)",
        y = "Interspecific (NN) Distance (%)") +
  geom_abline(intercept = 0, lty = 2)

plot_barcode_gap_scatterplot

```

## Figure 1
```{r}
fig1 <- 
ggdraw() + 
  
  # Draw main plot
  draw_plot(plot_family_count) +
  
  # Draw inset 1
  draw_plot(plot_distance_histogram +
              labs(fill = NULL) +
              theme(axis.text = element_text(size = 6),
                    axis.title = element_text(size = 7),
                    legend.title = element_text(size = 7),
                    legend.text = element_text(size = 7),
                    legend.key.size = unit(0.25, "cm"),
                    legend.position = "top"
              ),
            x = 0.12,
            y = 0.65,
            width=0.45, 
            height=0.35) +
  
  # Draw inset 2
  draw_plot(plot_barcode_gap_histogram +
              theme(axis.text = element_text(size = 6),
                    axis.title = element_text(size = 7),
                    legend.title = element_text(size = 7),
                    legend.text = element_text(size = 7),
                    legend.key.size = unit(0.25, "cm"),
                    legend.position = "top"
              ),
            x = 0.55,
            y = 0.65,
            width=0.45, 
            height=0.35) +
  
  # Draw inset 3
  draw_plot(plot_barcode_gap_scatterplot +
              theme(axis.text = element_text(size = 6),
                    axis.title = element_text(size = 7),
                    legend.title = element_text(size = 7),
                    legend.text = element_text(size = 7),
                    legend.key.size = unit(0.25, "cm"),
                    legend.position = "top"
              ),
            x = 0.55,
            y = 0.35,
            width=0.45, 
            height=0.25
            ) +   
  
  draw_plot_label(
    c("A", "B", "C", "D"),
    c(0,0.15, 0.55, 0.55),
    c(1, 1, 1, 0.65),
    size = c(11, 11, 11, 11)
    )

fig1

ggsave(fig1,
       file = paste0(outdir,
                     "fig1_data-description.png"),
       height = 6.5,
       width = 7,
       units = "in",
       dpi = 350
)

```



# Phylogenetics
Phylogenetic analysis based on R. Follows after the analysis pipeline on [Ph-IRES](https://github.com/Ph-IRES/workshop_data-analysis/tree/main/tutorial_phylogeny_analysis).

## Before you start

This pipeline assumes that the sequence has already been aligned. As much as possible, gaps/indels are removed from the alignment so as to avoid errors downstream. Furthermore, ambiguous sequences should have been resolved prior to uploading the data in R.

There are freely-available software that can guide in the upstream processes (e.g., sequence truncation and clean-up, alignment). @Hall2018 is a good reference for sequence alignment, as well as other topics involving phylogenetics.

## Libraries
```{r library}
library(pacman)

# Load packages
pacman::p_load(
  
  # Wrangling, Miscellaneous
  tidyverse,
  janitor,

  # Plotting
  ggpubr, 
  ggtree, 
  paletteer, 
  patchwork, 
  scales, 
  
  # Phylogenetics and Population Genetics Packages
  ade4, 
  adegenet, 
  ape, 
  bioseq,
  haplotypes, 
  pegas, 
  phangorn, 
  msa
)

```

## Modify labels

Change the labels of the fasta file based on the information from the metadata file.

```{r change_aln-label}
# Select only the columns needed
phylo_aln <- data_bkp %>% 
  mutate (mod_label = paste (taxa_id, sample_id, bin, sep = "|")) %>% 
  select (mod_label, sequence) %>% 
  as_DNAbin (labels = mod_label,
             sequences = sequence) %>% 
  as.matrix
```


# Collapse Haplotypes

```{r collapse-haplotypes}
haplo_list <- phylo_aln %>% 
  as.dna %>% 
  haplotypes::haplotype(., indels="sic")

# Haplotype report
sink (paste(outdir, "hapCollapse_report.txt"))
print (haplo_list)
sink()
  
# Set filters
filter <- map(haplo_list@haplist,
    ~.x[1]) %>% 
  plyr::ldply() %>% 
  pull (V1)

# Collapse haplotypes based on filter
phylo_haplo_aln <- phylo_aln[filter]

```

# Model Test

Determine the best evolutionary model for the given dataset using Maximum Likelihood.

```{r ML Model Selection}
# Conver collapsed alignment into a PhyDat object
pd<-as.phyDat(phylo_haplo_aln)

##test models, select lowest AIC
mt <- modelTest(pd, multicore = TRUE, mc.cores= 6)
View(mt)

# Best model: GTR + G(4) + I
```

# NJ Tree
```{r}
pd < -as.phyDat(phylo_aln)

nj.fun <- function (x) {
  x %>%   
    
  # Calculate genetic distance
  dist.dna(model = "K80",
           pairwise.deletion = T) %>% 
  
  # Draw NJ Tree
  NJ()
}

treeNJ <- nj.fun(phylo_aln)
  
bootstrap.pml (pd, bs=100)

```




# Maximum Likelihood

We now infer the phylogeny using maximum likelihood, and then visualize the resulting phylogenetic tree.

## Model Initialization

Initialize the model fit for the dataset.

```{r Initial ModelFit}
##initial fit, use best fit model from prev chunk
fit <- as.pml(mt)
fit
```

## Bootstrapping

Optimize the fit of the model and perform bootstrapping.

```{r Optimize Fit and Bootstrap}
##MAKE SURE TO CHANGE OPTIONS DEPENDING ON YOUR SELECTED MODEL.  
#?optim.pml gives guidance on how to set optBf and optQ
#optInv and optGamma should be set depending on whether your model includes +I and/or +G parameters

fit.opt <- optim.pml(fit, 
                     optBf=TRUE, 
                     optQ=TRUE, 
                     optInv=TRUE, 
                     optGamma=TRUE,
                     rearrangement = "NNI", 
                     control = pml.control(trace = 1)
                     )
fit.opt

#main= sets the plot title
plot(fit.opt, 
     main="GTR+G(4)+I")

#bootstrap model
bs <- bootstrap.pml(fit.opt, 
                    bs=1000, 
                    optNni=TRUE, 
                    multicore=TRUE, 
                    mc.cores = 6)

#Save bs in case as an R data file
# write_rds(bs,
#           file = paste0(outdir, "bootstrap.tre")
#           )

# In case you want to load the bs file:
# bs <- readRDS(paste0(outdir, "bootstrap.tre"))
          
```

## Initial Visualization

Look at the initial plot of the tree.

```{r}
tree<-plotBS(fit.opt$tree, bs, p = 50, digits = 0, type="p", method="FBP")

#digits doesn't seem to work in plotBS, so this is a workaround
tree$node.label<-round(tree$node.label, digits=0)

# plotBS also saves the tree and the bootstrap values in a phylo object, which we will use for visualization using ggtree.
```

```{r}
# Get a basic view of the tree
ggtree (tree,
        ladderize = T) +
  xlim (NA, 1) +
  geom_tiplab(geom="text", fontface=3, size=2.5, hjust = -0.025) +
  geom_text(aes(label=node), hjust=0.5, size = 2)


```

```{r}
# Set outgroup node
out.node <- 45
tree.root <- root(tree, node=out.node)

# Review tree after rooting
(treeplot <- 
ggtree (tree.root,
        ladderize = T) +
  xlim (NA, 1) +
  geom_tiplab(geom="text", fontface=3, size=3.5, hjust = -0.025) +
  geom_nodelab(size=3, hjust = 1.25, vjust=-0.2))

```

```{r modify and manipulate trees}

# Check node numbers
treeplot + geom_text(aes(label=node), hjust=0.5, size = 2)


#hilight nodes
treeplot_2 <- 
treeplot + 
  geom_hilight(node = 1, fill="darkred", extend = 0.55, alpha=0.25) +
  geom_treescale(width = 0.1, fontsize = 3, offset=-1)


```

## Exporting phylogeny

```{r tree tools}
##export tree in Newick format
write.tree(tree, file=paste0(outdir, "newick.tree"))

##export tree if png format
ggsave (treeplot_2,
        file = paste0(outdir, "phylo_tree.png"),
        width = 5,
        height = 7,
        units = "in")



```

